sklearn/manifold/tests/test_t_sne.py

```diff
@@ -879,6 +897,93 @@ def test_tsne_with_different_distance_metrics():
         assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
 
 
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+@ignore_warnings(category=FutureWarning)
+def test_tsne_with_legacy_euclidean_squaring(method):
+    """Make sure that TSNE preserves legacy Euclidean squaring behavior
+
+    Note: If 'legacy' is deprecated when the default for square_distance
+    is set to True, then this test can be removed, and 'euclidean' can
+    be included in `test_tsne_with_different_distance_metrics`."""
+    random_state = check_random_state(0)
+    n_components_original = 3
+    n_components_embedding = 2
+    X = random_state.randn(50, n_components_original).astype(np.float32)
+    X_transformed_tsne = TSNE(
+        metric='euclidean', n_components=n_components_embedding,
+        random_state=0, square_distance="legacy",
+        method=method).fit_transform(X)
+    X_precomputed = pairwise_distances(X, metric='euclidean', squared=True)
+    X_transformed_tsne_precomputed = TSNE(
+        metric='precomputed', n_components=n_components_embedding,
+        random_state=0, square_distance="legacy",
+        method=method).fit_transform(X_precomputed)
+    assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
```

Comment:
I believe I've uncovered an existing issue with this test. This test passes with `'exact'`, but fails with `'barnes_hut'`.

I think the reason why the `'barnes_hut'` test fails is because it uses `**=2` to square the distances, [here](https://github.com/scikit-learn/scikit-learn/blob/5edf203d6ee8a168c127b7bff7410ac83b497a30/sklearn/manifold/_t_sne.py#L809). But, the precomputed distance in this test uses `squared=True`.

To verify this is the case, I've written [a second test](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R945) which intentionally fails. It shows that TSNE produces different results depending on which squaring method is used (`**=2` or `squared=True`).